Get rid of package.json generation and yeoman generator

akosyakov opened the pull request.

akosyakov commented:
Just to clarify: we do it to streamline the setup of Theia itself as well as the setup for extension developers and simplify the dynamic extension system.

An extension developer will be able to create package.json using standard tools, like `npm install`, list extension packages and add `@theia/cli` as a dev dependency:
https://github.com/theia-ide/theia/blob/44f764c9f195364ace851e88772c6302adef06b1/examples/browser/package.json#L1-L35
After installing such package he will use `theia` cli tool to manage an application, see npm scripts above. So such command like `npx theia browser --port=3001 --root-dir=/home/user/mydir` or via npm script `yarn start  --port=3001 --root-dir=/home/user/mydir` would be possible.

For the Theia setup, we won't need the package.json generation phase anymore that allows removing some build scripts, simplify root package.json, yarn workspaces.

hexa00 commented:
What is not supported by node 6 in this ?

akosyakov commented:
@hexa00 async/await with pure js, bogus cluster module with node 6

akosyakov commented:
LTS of node 8 should start in October: https://github.com/nodejs/Release#release-schedule1

akosyakov commented:
I would like to merge if no objections and go on with the dynamic extension system.

svenefftinge commented on dev-packages/cli/package.json:
```diff
@@ -0,0 +1,48 @@
+    "url-loader": "^0.5.8",
+    "webpack": "^2.2.1",
+    "yo": "^2.0.0",
+    "generator-theia": "^0.1.1"
```
`yargs`seems to be missing

svenefftinge commented:
There are still some extension.package.json

akosyakov commented:
@svenefftinge fixed your comments

hexa00 commented on package.json:
```diff
@@ -3,34 +3,53 @@
     "test": "yarn run test:theia && yarn run test:electron && yarn run test:browser",
-    "test:theia": "node scripts/lerna run test --scope \"@theia/!(example-)*\" --parallel && yarn test:theia:slow",
-    "test:theia:slow": "node scripts/lerna run test:slow --scope \"@theia/terminal\" --parallel",
+    "test:theia": "node scripts/lerna run test --scope \"@theia/!(example-)*\" --parallel",
```
why did you remove test:theia:slow?

akosyakov replied:
do you run it?

akosyakov replied:
I kept them for projects, e.g. `npx lerna run test:slow --scope @theia/terminal`

hexa00 replied:
Yes it was run by CI by default, I would like to keep it that way

akosyakov replied:
it is still run by CI, each package has `test` script that runs `test:fast` + `test:slow`

hexa00 replied:
isn't that failing because many pacakge do not have slow tests ?

akosyakov replied:
I would like to simplify package.json when it is possible and remove scripts which are just there without anybody calling them regularly

akosyakov replied:
it is never failed if there are no tests, dummy files because of test coverage

hexa00 replied:
Yes the dummy files work for the test:fast pattern but not for the :slow

akosyakov replied:
it was the same before this PR should be tackled differently.

hexa00 replied:
No before this PR I explicitly selection that test:slow should be run on package/terminal only , otherwise the test run would fail

hexa00 replied:
See: the scope here:  `test:theia:slow": "node scripts/lerna run test:slow --scope \"@theia/terminal\" --parallel",`

hexa00 commented:
I get a bunch of warnings is that normal?
```
warning Pattern ["gulp-util@latest"] is trying to unpack in the same destination "/home/eantotr/.cache/yarn/v1/npm-gulp-util-3.0.8-0054e1e744502e27c04c187c3ecc505dd54bbb4f" as pattern ["gulp-util@^3.0.0","gulp-util@^3.0.1"]. This could result in a non deterministic behavior, skipping.
warning Pattern ["request@latest"] is trying to unpack in the same destination "/home/eantotr/.cache/yarn/v1/npm-request-2.82.0-2ba8a92cd7ac45660ea2b10a53ae67cd247516ea" as pattern ["request@^2.81.0","request@^2.72.0","request@2","request@^2.74.0","request@^2.45.0"]. This could result in a non deterministic behavior, skipping.
warning Pattern ["through@latest"] is trying to unpack in the same destination "/home/eantotr/.cache/yarn/v1/npm-through-2.3.8-0dd4c9ffaabc357960b1b724115d7e0e86a2e1f5" as pattern ["through@^2.3.6","through@^2.3.4","through@>=2.2.7 <3","through@2","through@~2.3.8","through@~2.3.4"]. This could result in a non deterministic behavior, skipping.
warning vscode-base-languageclient@0.0.1-alpha.3: The engine "vscode" appears to be invalid.
info fsevents@1.1.2: The platform "linux" is incompatible with this module.
info "fsevents@1.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning "font-awesome-webpack@0.0.5-beta.2" has incorrect peer dependency "font-awesome@>=4.3.0".
warning "url-loader@0.5.9" has incorrect peer dependency "file-loader@*".
warning "less-loader@2.2.3" has incorrect peer dependency "less@^2.3.1".
warning "ajv-keywords@1.5.1" has incorrect peer dependency "ajv@>=4.10.0".
warning The case-insensitive file /home/eantotr/sync/work/theia/node_modules/nyc/node_modules shouldn't be copied twice in one bulk copy
warning The case-insensitive file /home/eantotr/sync/work/theia/node_modules/nyc/node_modules/ansi-regex shouldn't be copied twice in one bulk copy
warning The case-insensitive file /home/eantotr/sync/work/theia/node_modules/nyc/node_modules/camelcase shouldn't be copied twice in one bulk copy


```

akosyakov commented:
It does not seem to harm, but we should look into them for sure, could be done separately.

hexa00 commented:
```
theia/process:   6 passing (4s)
@theia/process:   1 failing
@theia/process:   1) RawProcess test pipe stderr stream:
@theia/process:      Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
@theia/process:   

@theia/editor: Warning: Could not find any test files matching pattern: ./src/**/*.slow-spec.ts
@theia/editor: No test files found
 
etc...
```
Build returns 0

akosyakov commented:
@hexa00 Could you try it on master? Seems that some tests are bogus. Don't think that is related somehow to this PR.

akosyakov commented:
Nope, you are right, the timeout should fail the build, cannot find it in the log

hexa00 commented:
@akosyakov  ??? When I get an eror like that on maste the test run fails

hexa00 commented:
The @theia/editor: Warning: Could not find any test files matching pattern: ./src/**/*.slow-spec
also  failed the build before

akosyakov commented:
@hexa00 you are right

akosyakov commented:
@hexa00 What is with the test coverage if one runs nyc + mocha twice? Does it get merged? Or slow coverage overrides fast tests coverage?

akosyakov commented:
@hexa00 I've merged `test:fast` and `test:slow` in single `test` and adjusted pattern to cover all `*.*spec.ts` files that there is no special handling for the terminal package and everything is covered
https://github.com/theia-ide/theia/blob/e36aa557bf687ece15a95b44c64e671b57445ea8/dev-packages/ext-scripts/package.json#L26-L27

hexa00 commented:
I think the coverage gets merged.
OK for the merged tests if we can separate it when need be.

akosyakov commented:
@hexa00 yeah, it would be a bit messy with dummy spec files, we should look for the better way generally

hexa00 commented:
@akosyakov  Yes must be a way to treat warnings as warnings rather than errors

hexa00 commented:
Did you test the yarn.lock ?

akosyakov commented:
@hexa00 I fixed the propagation of the error code broken tests should fail the build now

hexa00 commented:
Testing that now

akosyakov commented:
@hexa00 nope, but should work, now lerna does not call install anymore, it is completely done by yarn

hexa00 commented:
Works fine , thanks! I need to increase that timeout

hexa00 commented:
Tested yarn.lock seems to work yeee!

hexa00 commented:
The log output is also so much easier to see!

akosyakov commented:
I've merged theia-generator into the cli:
- there are no dependencies on yo and yeoman-generator anymore, it makes windows builds stable
- there are no yo configs, modules are discovered with `require.replace` after installing
- webpack configs are regenerated each time and added to git ignore
- moved all dev dependencies to the root package, new should be added there as well

Please review.
